### ISSUE TITLE
Windows threads are using C run-time library incompatible CreateThread

### ISSUE THREAD

**Issue by Rantanen** [grpc/grpc#18509]
### What version of gRPC and what language are you using?
Git HEAD
 
### What operating system (Linux, Windows,...) and version?
Windows
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
MSVC / C++ 
 
### What did you do?

I've got a Windows gRPC server application and was looking into what kind of threads gRPC is using for synchronous service implementation. I dug my way into `thd_windows.cc` in gRPC core and found the following line:

https://github.com/grpc/grpc/blob/39a35fdb78d533e380b04886aa1172d61ad3445d/src/core/lib/gprpp/thd_windows.cc#L78

This supposedly calls the low level [`CreateThread`](https://docs.microsoft.com/en-us/windows/desktop/api/processthreadsapi/nf-processthreadsapi-createthread) WinAPI function? The documentation for that function states the following:

> A thread in an executable that calls the C run-time library (CRT) should use the _beginthreadex and _endthreadex functions for thread management rather than CreateThread and ExitThread; this requires the use of the multithreaded version of the CRT. If a thread created using CreateThread calls the CRT, the CRT may terminate the process in low-memory conditions.

Is this something that was considered in the gRPC implementation? I haven't seen any real issues cropping up due to this in my project so everything could be okay. I also have a feeling majority of gRPC production use happens on POSIX systems so the usage of CreateThread on Windows might not be as battle tested as the POSIX side.

**Comment by nicolasnoble** [comment]
Technically, as a library, using `_beginthreadex` forces the usage of which crt library to use down to the application. Using `CreateThread` instead which is lower level ensures that the application stays in control of their dependencies.

The point of the difference between the two is that if one is fully using the Microsoft crt, and relies on its behavior to be libc-style, such as reading `errno` after running a libc-style function, then the `_beginthreadex` wrapper is essential to ensure that `errno` is properly updated to the current thread.

As a comparison point, if you're using FreeRTOS for instance, it has this mechanism of interaction with newlib or glibc where there's a global specific to these libraries called `_impure_ptr` that's supposed to hold all of these reentrancy issues.

Long story short, the lower level grpc codebase shouldn't use any of these, as we're only supposed to be making low level calls, skipping the crt completely, and the threading here is concealed within its own context. If these are leaking to user-mode and calling back into user-mode, this would've been an issue, but this shouldn't be the case.

**Comment by Rantanen** [comment]
Ah, good to know. In that case I must have taken a wrong turn somewhere else. The answer I was looking for is the threads used to execute the requests in the synchronous server implementation. For async servers the threads are created by the user code, but with synchronous server some part of the gRPC code creates the threads.

My first guess would have been std::thread, but the only references that I seemed to find to std::thread in gRPC code base were in tests. I might have missed something though.

**Comment by Rantanen** [comment]
I'm still looking at this.

The C++ `Server` has a `SyncRequestThreadManager` inheriting from `ThreadManager`. According to the comment this is the mechanism for invoking the sync RPC handlers:

https://github.com/grpc/grpc/blob/59ffdd9929667d8d662d4bede6a21d4b3e35a93a/src/cpp/server/server_cc.cc#L630-L633

The `WorkerThread`s in the `ThreadManager` use `grpc_core::Thread`s:

https://github.com/grpc/grpc/blob/59ffdd9929667d8d662d4bede6a21d4b3e35a93a/src/cpp/thread_manager/thread_manager.cc#L30-L39

That `grpc_core::Thread` constructor on Windows uses `ThreadInternalsWindows` as the thread implementation:

https://github.com/grpc/grpc/blob/59ffdd9929667d8d662d4bede6a21d4b3e35a93a/src/core/lib/gprpp/thd_windows.cc#L144-L150

And the `ThreadInternalsWindows` creates the threads using `CreateThread`:

https://github.com/grpc/grpc/blob/59ffdd9929667d8d662d4bede6a21d4b3e35a93a/src/core/lib/gprpp/thd_windows.cc#L57-L78

As a result the user code in the sync RPC handlers ends up executed on top of threads that do not have the C run-time libs initialized on them?

I could also verify this in the debugger. The bottom frame in a sync RPC method (right above kernel32.dll) points to the CreateThread call in ThreadInternalsWindows.

> If these are leaking to user-mode and calling back into user-mode, this would've been an issue, but this shouldn't be the case.

Given that statement seems to contradict the code path shown above, I'll reopen the issue for now.

However I also tried calling stdlib `rand()` from a gRPC sync handler and that succeeded. If I followed the ucrt sources contained with Windows SDKs properly, `rand()` should attempt to read C run-time per thread data that is explicitly initialized by `_beginthread`. If this data wasn't initialized, rand() should abort the process - however it didn't. So it seems "things just work", but I have no idea why.

I'm suspecting this issue can still be just closed, however I'd sleep more peacefully if someone else confirmed that as well. :)

**Comment by nicolasnoble** [comment]
This is something probably for @vjpai to look at :)
